hasktorch/src/Torch/Data/Pipeline.hs

```diff
@@ -0,0 +1,112 @@
+                           , L.FoldM(..)
+                           , Dataset(..)
+                           , DatasetMock(..)
+                           , ConcurrentDataset(..)
+                           ) where
+
+import           Control.Arrow (first)
+import           Control.Concurrent.Async
+import qualified Control.Foldl as L
+import           Control.Monad
+import           Pipes
+import           Pipes.Concurrent
+import qualified Pipes.Prelude as P
+
+
+type Iter = Int
+type WorkerId = Int
+
+data DatasetMock m tensor = DatasetMock { getBatchMock :: Int -> m tensor
+                                        , numItersMock :: Iter
+                                        }
+
+class Dataset dataset batch  where
+  getBatch :: dataset -> Iter -> IO batch
+  numIters :: dataset -> Int
+
+class Dataset dataset batch => ConcurrentDataset dataset batch where
+  getBatchConcurrently :: WorkerId -> dataset -> Iter -> IO batch
+
+data RunBatch = Final | KeepTrain  deriving (Eq, Show)
+
+takeBatch :: MonadIO m => Input (batch, RunBatch) -> Producer batch m ()  
+takeBatch input = fromInput input >-> P.takeWhile ((/=) Final . snd) >-> P.map fst
+
+readBatchesConcurrently :: forall dataset batch m .
+  (MonadIO m, ConcurrentDataset dataset batch) => Int -> dataset -> Output (batch, RunBatch)  -> Effect m () 
+readBatchesConcurrently workerId dataset transformBox = 
+  for (each [1..numIters @dataset @batch dataset ]) (\iter -> yieldBatch iter >-> toOutput transformBox)
+    where yieldBatch iter =
+             (liftIO $ getBatchConcurrently workerId dataset iter) >>= yield . (, runBatch iter)
+          runBatch iter = if numIters @dataset @batch dataset == iter then Final else KeepTrain
+
+readBatches :: forall dataset batch m.
+  (MonadIO m, Dataset dataset batch) => dataset -> Output (batch, RunBatch)  -> Effect m () 
+readBatches dataset outputBox = 
+  for (each [1..numIters @dataset @batch dataset]) (\iter -> yieldBatch iter >-> toOutput outputBox)
+    where yieldBatch iter = (liftIO $ getBatch dataset iter) >>= yield . (, runBatch iter)
+          runBatch iter = if numIters @dataset @batch dataset == iter then Final else KeepTrain
+
+
+runTransforms :: MonadIO m => (tensor -> tensor') -> Input (tensor, RunBatch) -> Output (tensor', RunBatch) -> Effect m ()
+runTransforms transforms transformBox trainBox = fromInput transformBox >->  P.map (first transforms) >-> toOutput trainBox
+
+makeFoldWithTransform :: (MonadIO m, Dataset dataset batch)  
+  => (batch -> batch')
+  -> dataset
+  -> IO (L.FoldM m batch' b -> m b)
+makeFoldWithTransform transforms dataset = do
+          -- TODO: we can allow different buffer sizes
+          -- which would be necessary for data echoing
+            (toTransformBox, fromTransformBox, sealTransform) <- spawn' (bounded 1)
+            (toBatches, fromBatches, sealBatch) <- spawn' (bounded 1)
+            async $ runEffect $ forever $ readBatches dataset toTransformBox 
```

Comment:
hm, do you need `forever` here?

Reply:
Yes, since `readBatches` only yields data for 1 epoch, `forever` makes it yield an indefinite number of epochs.